### PR TITLE
refactor(Processor): remove unnecessary json.Marshalling while making common metadata from singular event

### PR DESCRIPTION
**Pull request by Sidddddarth**
## Description of the change

> Removing `json.Marshal` from `makeCommonMetadataFromSingularEvent`.
> Left the function for now to facilitate a benchmark test.

with Marshal:
```Running tool: /usr/local/go/bin/go test -benchmem -run=^$ -bench ^Benchmark_makeCommonMetadataFromSingularEvent_jsonMarshalling$ github.com/rudderlabs/rudder-server/processor

DB_DSN: postgres://rudder:password@localhost:55021/jobsdb?sslmode=disable
goos: darwin
goarch: arm64
pkg: github.com/rudderlabs/rudder-server/processor
Benchmark_makeCommonMetadataFromSingularEvent_jsonMarshalling-8   	  105403	     10408 ns/op	    4960 B/op	      70 allocs/op
PASS
ok  	github.com/rudderlabs/rudder-server/processor	9.073s
```
without:
```
Running tool: /usr/local/go/bin/go test -benchmem -run=^$ -bench ^Benchmark_makeCommonMetadataFromSingularEvent_withoutMarshalling$ github.com/rudderlabs/rudder-server/processor

DB_DSN: postgres://rudder:password@localhost:55022/jobsdb?sslmode=disable
goos: darwin
goarch: arm64
pkg: github.com/rudderlabs/rudder-server/processor
Benchmark_makeCommonMetadataFromSingularEvent_withoutMarshalling-8   	 1519579	       759.0 ns/op	     544 B/op	       3 allocs/op
PASS
ok  	github.com/rudderlabs/rudder-server/processor	6.907s
```

## Notion Link

> [notion](https://www.notion.so/rudderstacks/Remove-marshalling-from-makeCommonMetadataFromSingularEvent-a1f78573d37b46579b820a1d05f7c30f)

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added unit tests for the code
- [ ] I have made corresponding changes to the documentation

## Security

- [x] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

### CODE REVIEW

**Review by cisse21 — APPROVED** [review]

File: processor/processor.go [hunk 1]
```diff
@@ -700,6 +700,33 @@ func makeCommonMetadataFromSingularEvent(singularEvent types.SingularEventT, bat
...
 
+func makeCommonMetadataFromSingularEventWithoutMarshalling(singularEvent types.SingularEventT, batchEvent *jobsdb.JobT, receivedAt time.Time, source backendconfig.SourceT) *transformer.MetadataT {
+	commonMetadata := transformer.MetadataT{}
+	commonMetadata.SourceID = gjson.GetBytes(batchEvent.Parameters, "source_id").Str
+	commonMetadata.WorkspaceID = source.WorkspaceID
+	commonMetadata.Namespace = config.GetKubeNamespace()
+	commonMetadata.InstanceID = config.GetInstanceID()
+	commonMetadata.RudderID = batchEvent.UserID
+	commonMetadata.JobID = batchEvent.JobID
+	commonMetadata.MessageID = singularEvent["messageId"].(string)
```
Comment (lvrach) [inline processor/processor.go:711]: I am bit sceptical regarding the unchecked casting.

File: utils/misc/misc.go [hunk 2]
```diff
@@ -1394,3 +1393,17 @@ func ReverseInt(s []int) []int {
...
+// lookup map recursively and return value
+func MapLookup(mapToLookup map[string]interface{}, defaultVal interface{}, keys ...string) interface{} {
+	if len(keys) == 0 {
+		return defaultVal
+	}
+	if val, ok := mapToLookup[keys[0]]; ok {
+		if len(keys) == 1 {
+			return val
+		}
+		return MapLookup(val.(map[string]interface{}), defaultVal, keys[1:]...)
```
Comment (atzoum) [inline utils/misc/misc.go:1406]: A type assertion would be valuable
```suggestion
		nextMap, ok := val.(map[string]interface{})
		if !ok {
			return defaultVal
		}
		return MapLookup(nextMap, defaultVal, keys[1:]...)
```

File: processor/processor.go [hunk 3]
```diff
@@ -700,6 +700,33 @@ func makeCommonMetadataFromSingularEvent(singularEvent types.SingularEventT, bat
...
+	commonMetadata.Namespace = config.GetKubeNamespace()
+	commonMetadata.InstanceID = config.GetInstanceID()
+	commonMetadata.RudderID = batchEvent.UserID
+	commonMetadata.JobID = batchEvent.JobID
+	commonMetadata.MessageID = misc.GetStringifiedData(singularEvent["messageId"])
+	commonMetadata.ReceivedAt = receivedAt.Format(misc.RFC3339Milli)
+	commonMetadata.SourceType = source.SourceDefinition.Name
+	commonMetadata.SourceCategory = source.SourceDefinition.Category
+
+	commonMetadata.SourceBatchID = misc.MapLookup(singularEvent, "", "context", "sources", "batch_id").(string)
```
Comment (atzoum) [inline processor/processor.go:716]: all `.(string)` type assertions below can theoretically fail. With this small trick you are safe and you no longer need to pass "" as a default value to `MapLookup` (can remove the parameter and return `nil` instead)
```suggestion
	commonMetadata.SourceBatchID, _ = misc.MapLookup(singularEvent, "", "context", "sources", "batch_id").(string)
```

File: processor/processor.go [hunk 4]
```diff
@@ -700,6 +700,33 @@ func makeCommonMetadataFromSingularEvent(singularEvent types.SingularEventT, bat
 	return &commonMetadata
 }
 
+func makeCommonMetadataFromSingularEventWithoutMarshalling(singularEvent types.SingularEventT, batchEvent *jobsdb.JobT, receivedAt time.Time, source backendconfig.SourceT) *transformer.MetadataT {
```
Comment (atzoum) [inline processor/processor.go:703]: Do we need 2 methods? Why not updating the original method?

**Review by atzoum — CHANGES_REQUESTED** [review]

**Comment by atzoum** [comment]
Please update the pull request's title according to our [new policy](https://www.notion.so/rudderstacks/Pull-Requests-40a4c6bd7a5e4387ba9029bab297c9e3)

File: processor/processor.go (hunk 4 above)
Comment (Sidddddarth) [reply processor/processor.go:703]: No we don't need it.
Just left it for now for facilitating the benchmark test.
Will be removed before marking ready for review.

File: processor/processor.go (hunk 3 above)
Comment (Sidddddarth) [reply processor/processor.go:716]: Nice. Something felt amiss. This makes sense.

File: utils/misc/misc.go (hunk 2 above)
Comment (Sidddddarth) [reply utils/misc/misc.go:1406]: Nice. Done.
This function brought to you by github copilot!

**Review by chandumlg — APPROVED** [review]

**Review by atzoum — APPROVED** [review]

**Pull request merged**